File: doc-examples/mongo-reactive-example-kotlin/src/main/kotlin/example/ProductRepository.kt

package example

import io.micronaut.data.annotation.*
import io.micronaut.data.mongodb.annotation.MongoRepository
import io.micronaut.data.repository.CrudRepository
import org.bson.types.ObjectId
import reactor.core.publisher.Mono
import java.util.concurrent.CompletableFuture
// tag::join[]
// tag::async[]
@MongoRepository
interface ProductRepository : CrudRepository<Product, ObjectId> {
// end::join[]
// end::async[]

    // tag::join[]
    @Join("manufacturer") // <1>
    fun list(): List<Product>
    // end::join[]

    // tag::async[]
    @Join("manufacturer")
    fun findByNameRegex(str: String): CompletableFuture<Product>

    fun countByManufacturerName(name: String?): CompletableFuture<Long>
    // end::async[]

    // tag::reactive[]
    @Join("manufacturer")
    fun queryByNameRegex(str: String): Mono<Product>

    fun countDistinctByManufacturerName(name: String?): Mono<Long>
    // end::reactive[]

// tag::join[]
// tag::async[]
}
// end::join[]
// end::async[]

File: doc-examples/mongo-example-python/src/main/python/example/ProductRepository.py
from java.util.concurrent import CompletableFuture
from micronaut.data.annotation import Join
from micronaut.data.mongodb.annotation import MongoRepository
from micronaut.data.repository import CrudRepository
from org.bson.types import ObjectId
from reactor.core.publisher import Mono

from example.Product import Product


# tag::join[]
# tag::async[]
@MongoRepository
class ProductRepository(CrudRepository[Product, ObjectId]):
    # end::join[]
    # end::async[]

    # tag::join[]
    @Join("manufacturer")  # <1>
    def list(self) -> list[Product]: ...
    # end::join[]

    # tag::async[]
    @Join("manufacturer")
    def findFirstByNameRegex(self, str: str) -> CompletableFuture[Product]: ...

    def countByManufacturerName(self, name: str) -> CompletableFuture[int]: ...
    # end::async[]

    # tag::reactive[]
    @Join("manufacturer")
    def queryFirstByNameRegex(self, str: str) -> Mono[Product]: ...

    def countDistinctByManufacturerName(self, name: str) -> Mono[int]: ...
    # end::reactive[]
